kotest-assertions/kotest-assertions-shared/src/commonMain/kotlin/io/kotest/assertions/eq/IterableEq.kt

```diff
@@ -62,45 +65,86 @@ object IterableEq : Eq<Iterable<*>> {
+      val detailErrorMessage = when {
+         actual is Set<*> || expected is Set<*> -> "Set can be compared only to Set\nCannot compare $tag"
+         actual is Collection || expected is Collection -> "Collection can be compared only to Collection\nCannot compare $tag"
+         else -> "Promiscuous iterators\nCannot compare $tag"
+      }
+      return failure(Expected(expected.print()), Actual(actual.print()), detailErrorMessage)
+   }
+
+   private const val unim = "Unsupported nesting iterator"
+
    private fun checkEquality(actual: Iterable<*>, expected: Iterable<*>, strictNumberEq: Boolean): Throwable? {
       var index = 0
       val iter1 = actual.iterator()
       val iter2 = expected.iterator()
       val elementDifferAtIndex = mutableListOf<Int>()
       var unexpectedElementAtIndex: Int? = null
       var missingElementAt: Int? = null
+      var nestedIteratorError: String? = null
+      var accrueDetails = true
       while (iter1.hasNext()) {
          val a = iter1.next()
          if (iter2.hasNext()) {
             val b = iter2.next()
-            val t = eq(a, b, strictNumberEq)
-            if (t != null) {
+            val t: Throwable? = when {
+               a?.equals(b) == true -> null
+               (a is Iterable<*>) && (a !is Collection<*>) && (a::class.isInstance(actual) || actual::class.isInstance(a)) -> {
+                  nestedIteratorError = "$unim $a within $iter1 (must use custom code)"
+                  accrueDetails = false
+                  failure(nestedIteratorError)
+               }
+               (b is Iterable<*>) && (b !is Collection<*>) && (b::class.isInstance(expected) || expected::class.isInstance(
+                  b
+               )) -> {
+                  nestedIteratorError = "$unim $b within $iter2 (must use custom code)"
+                  accrueDetails = false
+                  failure(nestedIteratorError)
+               }
+               else -> eq(a, b, strictNumberEq)?.let {
+                  if (it.message?.startsWith(unim) == true) {
+                     nestedIteratorError = it.message
+                     accrueDetails = false
+                     failure(nestedIteratorError!!)
+                  } else it
+               }
+            }
+            if (t != null && accrueDetails) {
                elementDifferAtIndex.add(index)
             }
-         } else {
-            unexpectedElementAtIndex = index
-         }
+         } else if (accrueDetails) unexpectedElementAtIndex = index else break
          index++
       }
       if (iter2.hasNext()) {
          missingElementAt = index
       }
       val detailErrorMessage = StringBuilder().apply {
          if (elementDifferAtIndex.isNotEmpty()) {
-            append("Element differ at index: ${elementDifferAtIndex.print().value}\n")
+            nestedIteratorError?.let {
```

Comment:
This seems redundant, if there is `nestedIteratorError` then at `line 143` then your returning error of `nestedIteratorError` this `detailErrorMessage` is getting thrown only when `nestedIteratorError` is not present.

Reply:
Yup!  Thanks, removed.